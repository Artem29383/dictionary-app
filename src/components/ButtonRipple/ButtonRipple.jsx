import React from 'react';
import PropTypes from 'prop-types';
import S from './ButtonRipple.styled';

const ButtonRipple = ({ children, className, onClick, dataClose }) => (
  <S.Button className={className} onClick={onClick} data-close={dataClose}>
    {children}
  </S.Button>
);

export default ButtonRipple;

ButtonRipple.propTypes = {
  children: PropTypes.string,
  className: PropTypes.string,
  onClick: PropTypes.func,
  dataClose: PropTypes.bool,
};

ButtonRipple.defaultProps = {
  dataClose: false,
};
