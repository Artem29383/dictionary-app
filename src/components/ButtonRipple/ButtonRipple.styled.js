import styled from 'styled-components/macro';
import { colors } from 'styles/constants';

export default {
  Button: styled.button`
    border: none;
    border-radius: 2px;
    padding: 0 18px;
    line-height: 2.5;
    font-size: 16px;
    text-transform: uppercase;
    cursor: pointer;
    display: block;
    color: ${colors.white};
    background-color: ${colors.dodjerBlue};
    box-shadow: 0 0 4px ${colors.dustyGray};
    background-position: center;
    transition: background 0.8s;

    &.circle {
      border-radius: 50%;
      height: 100%;
      width: 100%;
      line-height: 0;
    }

    &:hover {
      background: ${colors.pictonBlue}
        radial-gradient(circle, transparent 1%, ${colors.pictonBlue} 1%)
        center/15000%;
    }

    &.green {
      background-color: ${colors.cornHarvest};

      &:hover {
        background: ${colors.crete}
          radial-gradient(circle, transparent 1%, ${colors.crete} 1%)
          center/15000%;
      }

      &:active {
        background-color: ${colors.bitter};
        background-size: 100%;
        transition: background 0s;
      }
    }

    &.red {
      background-color: ${colors.vermillion};

      &:hover {
        background: ${colors.blazeOrange}
          radial-gradient(circle, transparent 1%, ${colors.blazeOrange} 1%)
          center/15000%;
      }

      &:active {
        background-color: ${colors.bitter};
        background-size: 100%;
        transition: background 0s;
      }
    }

    &:active {
      background-color: ${colors.malibu};
      background-size: 100%;
      transition: background 0s;
    }

    &.center {
      margin: 0 auto;
    }
  `,
};
