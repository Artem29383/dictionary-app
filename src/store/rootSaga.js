import { all } from '@redux-saga/core/effects';
import rootSagaAuth from 'models/user/saga';
import rootSagaDictionary from 'models/dictionary/saga';
import rootSagaTest from 'models/test/saga';

export function* rootSaga() {
  yield all([rootSagaAuth(), rootSagaDictionary(), rootSagaTest()]);
}
