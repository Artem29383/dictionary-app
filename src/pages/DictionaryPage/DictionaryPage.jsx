import React, { useEffect, useState } from 'react';
import Table from 'components/Table';
import orderBy from 'lodash.orderby';
import useAction from 'hooks/useAction';
import { GET_DICTIONARY } from 'models/dictionary/action';
import { setDictionary, setLoading } from 'models/dictionary/reducer';
import useSelector from 'hooks/useSelector';
import {
  getFilteredWords,
  getIdsSelector,
  getLoadingSelector,
  getWordsSelector,
} from 'models/dictionary/selectors';
import Loader from 'components/Loader';
import ButtonRipple from 'components/ButtonRipple';
import useToggle from 'hooks/useToggle';
import { normalized } from 'utils/normalized';
import { useInput } from 'hooks/useInput';
import TableSearch from 'components/TableSearch';
import { chunks } from 'utils/chunks';
import Paginate from 'components/Paginate';
import AddWordModal from 'pages/DictionaryPage/AddWordModal';
import { getUserSelector } from 'models/user/selectors';
import Portal from 'components/Portal';
import DropDown from 'components/DropDown';
import S from './DictionaryPage.styled';

const DictionaryPage = () => {
  const [user] = useSelector(getUserSelector);
  const getDictionary = useAction(GET_DICTIONARY);
  const setLoad = useAction(setLoading);
  const isLoading = useSelector(getLoadingSelector);
  const words = useSelector(getWordsSelector);
  const setDict = useAction(setDictionary);
  const [currentPage, setCurrentPage] = useState(0);
  const ids = useSelector(getIdsSelector);
  const [elemCountsOnPage, setElemCountsOnPage] = useState(10);
  const [sort, setSort] = useState('asc');
  const [value, setValue] = useInput('');
  const filteredIds = useSelector(getFilteredWords)(value);
  const [showModal, setShowModal] = useToggle(false);
  const [chunksArray, setChunksArray] = useState([]);
  const chunkArrayLength = chunksArray.length;

  const onSort = (sortCol, initSort) => {
    const sortType = sort === 'asc' ? 'desc' : 'asc';
    const displayData = orderBy(words, sortCol, initSort || sortType);
    const dataNormalized = normalized(displayData, 'dictionary');
    setDict({
      entities: dataNormalized.entities.dictionary,
      ids: dataNormalized.result,
    });
    setSort(initSort || sortType);
  };

  useEffect(() => {
    onSort('word', 'asc');
  }, [ids.length]);

  useEffect(() => {
    if (filteredIds.length === 0) {
      setLoad(true);
      getDictionary(user.login);
    }
  }, []);

  useEffect(() => {
    setChunksArray(chunks(filteredIds, elemCountsOnPage));
    setCurrentPage(0);
  }, [elemCountsOnPage, filteredIds, ids.length]);

  return (
    <S.Content>
      <TableSearch
        type="text"
        value={value}
        onChange={setValue}
        label="Поиск"
      />
      <DropDown
        className="dictionary"
        label="Элементов на странице:"
        value={String(elemCountsOnPage)}
        setValue={setElemCountsOnPage}
        options={[10, 50, 100]}
      />
      {isLoading ? (
        <Loader />
      ) : (
        <Table
          userId={user.id}
          login={user.login}
          sortType={sort}
          sort={onSort}
          words={words}
          ids={chunksArray[currentPage] || []}
        />
      )}
      {chunkArrayLength > 1 ? (
        <Paginate
          currentPage={currentPage + 1}
          chunkArrayLength={chunkArrayLength}
          setCurrentPage={setCurrentPage}
        />
      ) : null}
      <S.BtnPos>
        <ButtonRipple clickHandler={setShowModal} className="circle">
          ✚
        </ButtonRipple>
      </S.BtnPos>
      <Portal
        id="Modal"
        /* eslint-disable-next-line react/no-children-prop */
        children={
          showModal && (
            <AddWordModal
              id={user.id}
              login={user.login}
              toggle={setShowModal}
              isOpen={showModal}
            />
          )
        }
      />
    </S.Content>
  );
};

export default DictionaryPage;
