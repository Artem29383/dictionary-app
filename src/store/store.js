import { combineReducers } from 'redux';
import createSagaMiddleware from 'redux-saga';
import { configureStore } from '@reduxjs/toolkit';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import { createBrowserHistory } from 'history';
import authReducer from 'models/user/reducer';
import dictionaryReducer from 'models/dictionary/reducer';
import testReducer from 'models/test/reducer';
import { rootSaga } from './rootSaga';

export const history = createBrowserHistory();
const saga = createSagaMiddleware();
const isDev = process.env.NODE_ENV === 'development';

const reducer = combineReducers({
  auth: authReducer,
  dictionary: dictionaryReducer,
  test: testReducer,
  router: connectRouter(history),
});

const store = configureStore({
  reducer,
  middleware: [saga, routerMiddleware(history)],
  devTools: isDev,
});
saga.run(rootSaga);

export default store;
