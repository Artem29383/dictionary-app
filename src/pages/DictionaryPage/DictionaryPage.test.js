import { configure, shallow, mount, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import DictionaryPage from 'pages/DictionaryPage/DictionaryPage';
import React from 'react';
import { Provider } from 'react-redux';
import store from 'store/store';
import ButtonRipple from 'components/ButtonRipple';

configure({ adapter: new Adapter() });

describe('Testing Dictionary page', () => {
  let action;
  let wrapper;
  
  beforeEach(() => {
    action = jest.fn();
    wrapper = shallow(
      <Provider store={store}>
        <DictionaryPage />
      </Provider>
    );
  });
  
  test('Render', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  test('Render button', () => {
    const button = wrapper.find(<ButtonRipple />);
    button.toMatchSnapshot();
  })
});