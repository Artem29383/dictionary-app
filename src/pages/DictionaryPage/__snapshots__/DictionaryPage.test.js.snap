// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Dictionary page Render 1`] = `ShallowWrapper {}`;
